### PR TITLE
make log directory configurable during init

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

* the log directory configured in tedge config set logs.path
is now used in --init

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
#1070

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_agent/src/agent.rs [hunk 1]
```diff
@@ -225,10 +225,24 @@ impl SmAgent {
         })
     }
 
+    pub fn get_log_dir(&self) -> PathBuf {
```
Comment (matthiasbeyer) [inline crates/core/tedge_agent/src/agent.rs:228]:
```suggestion
    pub fn log_dir(&self) -> PathBuf {
```

Getter functions are normally called only like the variable they're accessing. So no `get_` prefix! :wink:

File: crates/core/tedge_agent/src/agent.rs [hunk 2]
```diff
@@ -225,10 +225,24 @@ impl SmAgent {
...
-    pub async fn init(&mut self, config_dir: PathBuf) -> Result<(), anyhow::Error> {
+    pub async fn init(
+        &mut self,
+        config_dir: PathBuf,
+        log_path: PathBuf,
+    ) -> Result<(), anyhow::Error> {
+        // `config_dir` by default is `/etc/tedge` (or whatever the user sets with --config-dir)
+        // `log_path` is by default `/var/log` (or whatever the user sets with tedge config set logs.path)
+        let cfg_dir = config_dir.as_path().display().to_string();
+        let log_path = log_path.as_path().display().to_string();
```
Comment (matthiasbeyer) [inline crates/core/tedge_agent/src/agent.rs:241]: Is this necessary? `PathBuf` can be made `Display`, so the interpolation should work without the full conversion to string!

```suggestion
        let cfg_dir = config_dir.display();
        let log_path = log_path.display();
```

You might want to rename the variables though.

File: crates/core/tedge_agent/src/main.rs [hunk 3]
```diff
@@ -51,8 +51,11 @@ async fn main() -> Result<(), anyhow::Error> {
         "tedge_agent",
         SmAgentConfig::try_new(tedge_config_location)?,
     )?;
+
+    let log_dir = agent.get_log_dir();
```
Comment (matthiasbeyer) [inline crates/core/tedge_agent/src/main.rs:55]:
```suggestion
    let log_dir = agent.log_dir();
```

:wink:

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Found some minor things. :wink:

**Comment by matthiasbeyer** [comment]
Ah, and commit messages should use imperative mood:

So instead of

```
Making log directory configurable during init
```

you'd write

```
Make log directory configurable during init
```

:wink:

**Comment by initard** [comment]
@matthiasbeyer thanks!

**Review by matthiasbeyer — APPROVED** [review]
Code-wise this looks good to me now! :laughing:

File: crates/core/tedge_agent/src/agent.rs [hunk 4]
```diff
@@ -225,10 +225,24 @@ impl SmAgent {
...
+    pub fn log_dir(&self) -> PathBuf {
+        self.config.log_dir.clone()
+    }
+
     #[instrument(skip(self), name = "sm-agent")]
-    pub async fn init(&mut self, config_dir: PathBuf) -> Result<(), anyhow::Error> {
+    pub async fn init(
+        &mut self,
+        config_dir: PathBuf,
+        log_path: PathBuf,
```
Comment (didier-wenzek) [inline crates/core/tedge_agent/src/agent.rs:236]: Is this new `log_path` argument necessary? `self.log_dir()` should make it. Isn't it?

File: crates/core/tedge_agent/src/agent.rs [hunk 5]
```diff
@@ -225,10 +225,24 @@ impl SmAgent {
...
+    ) -> Result<(), anyhow::Error> {
+        // `config_dir` by default is `/etc/tedge` (or whatever the user sets with --config-dir)
+        // `log_path` is by default `/var/log` (or whatever the user sets with tedge config set logs.path)
+        let config_dir = config_dir.display();
+        let log_path = log_path.display();
+        create_directory_with_user_group(&format!("{config_dir}/.agent"), "tedge", "tedge", 0o775)?;
+        create_directory_with_user_group(&format!("{log_path}/tedge"), "tedge", "tedge", 0o775)?;
         create_directory_with_user_group(
```
[the remark this answers is not in the document]
Comment (initard) [reply crates/core/tedge_agent/src/agent.rs:236]: oops, yes

File: crates/core/tedge_agent/src/agent.rs [hunk 6]
```diff
@@ -225,19 +225,11 @@ impl SmAgent {
...
-        &mut self,
-        config_dir: PathBuf,
-        log_path: PathBuf,
-    ) -> Result<(), anyhow::Error> {
+    pub async fn init(&mut self, config_dir: PathBuf) -> Result<(), anyhow::Error> {
         // `config_dir` by default is `/etc/tedge` (or whatever the user sets with --config-dir)
-        // `log_path` is by default `/var/log` (or whatever the user sets with tedge config set logs.path)
         let config_dir = config_dir.display();
+        let log_path = self.config.log_dir.clone();
         let log_path = log_path.display();
```
Comment (matthiasbeyer) [inline crates/core/tedge_agent/src/agent.rs:233]:
```suggestion
        let log_path = self.config.log_path.display();
```

I think the `clone()` call can be removed here, right?

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]

File: crates/core/tedge_agent/src/agent.rs (hunk 6 above)
Comment (initard) [reply crates/core/tedge_agent/src/agent.rs]: thanks :D

**Review by didier-wenzek — APPROVED** [review]
Approved

**Review by matthiasbeyer — APPROVED** [review]
There's one unrelated whitespace change, but that's not so important :smile: 

Please don't forget to squash the fixup commits!

File: crates/core/tedge_agent/src/main.rs [hunk 7]
```diff
@@ -51,6 +51,7 @@ async fn main() -> Result<(), anyhow::Error> {
         "tedge_agent",
         SmAgentConfig::try_new(tedge_config_location)?,
     )?;
+
```
Comment (albinsuresh) [inline crates/core/tedge_agent/src/main.rs:54]: Don't forget to revert that.

**Review by albinsuresh — APPROVED** [review]

**Comment by initard** [comment]
sorry whitespace police! @matthiasbeyer @albinsuresh

**Pull request merged**